mobile/src/main/java/org/openhab/habdroid/ui/OpenHABWidgetAdapter.java

```diff
@@ -59,918 +55,1054 @@
-        } else if (openHABWidget.getType().equals("Selection")) {
-            return TYPE_SELECTION;
-        } else if (openHABWidget.getType().equals("Setpoint")) {
-            return TYPE_SETPOINT;
-        } else if (openHABWidget.getType().equals("Chart")) {
-            return TYPE_CHART;
-        } else if (openHABWidget.getType().equals("Video")) {
-            if (openHABWidget.getEncoding() != null) {
-                if (openHABWidget.getEncoding().toLowerCase().equals("mjpeg")) {
-                    return TYPE_VIDEO_MJPEG;
+        }
+    }
+
+    public static class GenericViewHolder extends ViewHolder {
+        private final TextView mLabelView;
+        private final MySmartImageView mIconView;
+
+        GenericViewHolder(LayoutInflater inflater, ViewGroup parent,
+                MyAsyncHttpClient httpClient, ConnectionInfo connection) {
+            super(inflater, parent, R.layout.openhabwidgetlist_genericitem, httpClient, connection);
+            mLabelView = itemView.findViewById(R.id.widgetlabel);
+            mIconView = itemView.findViewById(R.id.widgetimage);
+        }
+
+        @Override
+        public void bind(OpenHABWidget widget) {
+            mLabelView.setText(widget.getLabel());
+            updateTextViewColor(mLabelView, widget.getLabelColor());
+            updateIcon(mIconView, widget);
+        }
+    }
+
+    public static class FrameViewHolder extends ViewHolder {
+        private final TextView mLabelView;
+
+        FrameViewHolder(LayoutInflater inflater, ViewGroup parent,
+                MyAsyncHttpClient httpClient, ConnectionInfo connection) {
+            super(inflater, parent, R.layout.openhabwidgetlist_frameitem, httpClient, connection);
+            mLabelView = itemView.findViewById(R.id.widgetlabel);
+            itemView.setClickable(false);
+        }
+
+        @Override
+        public void bind(OpenHABWidget widget) {
+            mLabelView.setText(widget.getLabel());
+            updateTextViewColor(mLabelView, widget.getValueColor());
+            // hide empty frames
+            itemView.setVisibility(widget.getLabel().isEmpty() ? View.GONE : View.VISIBLE);
+        }
+    }
+
+    public static class GroupViewHolder extends ViewHolder {
+        private final TextView mLabelView;
+        private final TextView mValueView;
+        private final MySmartImageView mIconView;
+
+        GroupViewHolder(LayoutInflater inflater, ViewGroup parent,
+                MyAsyncHttpClient httpClient, ConnectionInfo connection) {
+            super(inflater, parent, R.layout.openhabwidgetlist_groupitem, httpClient, connection);
+            mLabelView = itemView.findViewById(R.id.widgetlabel);
+            mValueView = itemView.findViewById(R.id.widgetvalue);
+            mIconView = itemView.findViewById(R.id.widgetimage);
+        }
+
+        @Override
+        public void bind(OpenHABWidget widget) {
+            String[] splitString = widget.getLabel().split("\\[|\\]");
```

Comment:
Consider moving `"\\[|\\]"` to a global variable. That's a simple regex, but a variable name could describe it.

Reply:
I actually thought about moving the split-label-to-label-and-value parsing to a common base class, but the handling is subtly different for each different type (fallback handling if there's less than 2 strings in the resulting array differs). I wonder whether that's intended or a slight bug which could be solved by using a base class?